Common/Extensions.cs

```diff
@@ -40,6 +42,36 @@ namespace QuantConnect
     /// </summary>
     public static class Extensions
     {
+        /// <summary>
+        /// Given a type will create a new instance using the parameterless constructor
+        /// and assert the type implements <see cref="IBaseData"/>
+        /// </summary>
+        /// <remarks>One of the objectives of this method is to normalize the creation of the
+        /// IBaseData instances while reducing code duplication</remarks>
+        public static IBaseData GetBaseDataInstance(this Type type)
+        {
+            var objectActivator = ObjectActivator.GetActivator(type);
+            if (objectActivator == null)
+            {
+                throw new ArgumentException($"Data type \'{type.Name}\' missing parameterless constructor " +
+                    $"E.g. public {type.Name}() {{ }}");
+            }
+
+            var instance = objectActivator.Invoke(new object[] { type });
+            if(instance == null)
+            {
+                // shouldn't happen but just in case...
+                throw new ArgumentException($"Failed to create instance of type \'{type.Name}\'");
+            }
+
+            var result = instance as IBaseData;
+            if (result == null)
```

Comment:
This type check can be performed statically via `typeof(IBaseData).IsAssignableFrom(type)` before invoking the `objectActivator` to save a few cycles and fail early.

Reply:
Good point - We expect 'instance' to be `IBaseData` in most cases so I went for 'as' versus 'IsAssignableFrom' since it is actually slightly cheaper - Check the following performance test:
```
        [Test]
        public void IsAssignableFromVsAsPerformanceTest()
        {
            Stopwatch stopwatch = new Stopwatch();
            object value = "This is a test";
            var type = value.GetType();
            stopwatch.Start();

            var result = IsAssignableFromTest(type);
            stopwatch.Stop();
            Console.WriteLine("IsAssignableFrom: {0} ms elapsed", stopwatch.ElapsedMilliseconds);
            Assert.IsTrue(result);

            stopwatch.Restart();
            result = AsTest(value);
            stopwatch.Stop();
            Assert.IsTrue(result);
            Console.WriteLine("AsOperator: {0} ms elapsed", stopwatch.ElapsedMilliseconds);
        }
        private bool IsAssignableFromTest(Type type, int loopCount = 1000000)
        {
            var isConvertible = false;
            for (int attempt = 0; attempt < loopCount; attempt++)
            {
                isConvertible |= typeof(IConvertible).IsAssignableFrom(type);
            }
            return isConvertible;
        }
        private bool AsTest(object value, int loopCount = 1000000)
        {
            var isConvertible = false;
            for (int attempt = 0; attempt < loopCount; attempt++)
            {
                isConvertible |= value as string != null;
            }
            return isConvertible;
        }
```

PS: I'll add a comment